### ISSUE TITLE
Choose a testing framework to begin building an automated test suite

### ISSUE THREAD

**Issue by tuxayo** [Coral-erm/Coral#114]
# TL;DR

We need tests in Coral. GUI tests are the most efficient way to **start** testing. Codeception is the best option for GUI testing.
# Motivations: why do we need a test suite? (skip if already convinced)
- detect regressions on the most critical use cases and features to ensure that they never break.
- detect regressions earlier to allow the developer that already has the context to fix them immediately. Instead of someone else much later who will spend more time to understand what went wrong.
- spend less time manually testing the most common use cases.
  - important when some of the features require dozens of clicks and more than 20 sec to be tested (setting up the DB in a particular state) which is tedious, error prone and can't be done after each change.
  - even more important considering the amount of refactoring work that must be done and that the lowest layers can impact all the features.
  - as the test suite will become more comprehensive the QA work before each release will stay constant or will decrease (depending on the test coverage). While without automated tests, it will increase thus making harder to release as time goes on.
- less risk of breaking something else when doing a hotfix.
- document the most common and some complex use cases for the developers. Useful when not knowing well the domain (ERMS) and Coral. It allows to integrate new devs easier and reassure them about not breaking everything without noticing it.
- easily test with multiple versions of PHP, MariaDB/MySQL and reduce  the cost of supporting a new DBMS. Over the years it will avoid some bad surprises (`mysql` API dropped in PHP 7, [default date values changes](https://github.com/Coral-erm/Coral/issues/50) in MariaDB/MySQL)
- make comprehensive and reproducible performance benchmarks
- as Coral will grow in features, complexity and users, the benefits of having a more and more comprehensive test suite will increase. It's already risky to touch the lower layers without breaking something far away from there that will only (hopefully) be detected in the QA phase before the next release (mysql→mysqli migration). Thus having to test everything manually is a barrier to refactoring. Refactoring is essential on the long run to improve the code quality and prevent the codebase to become "untouchable".
### database merge (between all modules) #65

It could potentially introduce regressions in every part of the app. If started now, this work would require a huge amount of manual testing of the main features at many stages of the refactoring process. Having them automated would allow to save a lot of time + fix regressions just after they were introduced instead of hours/days latter.
### deduplicate common code

For example in `*/admin/classes/` for each module we have many identical(or almost) classes.

Same problems as for merging the DB
### separate view from logic #78

There are some screens that are difficult/long to access and having automated tests would allow to easily check that the functionality works + taking a screenshot from the test would help to check that the page looks good.
### integrating quickly the finished work

We have a significant backlog of PRs to merge.
- Having tests on the most important features would give more confidence in merging.
  In the long term having a comprehensive test suite will allow us to move towards [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration) because we are already to some extent in the "integration hell" (without even doing big refactorings)
- Having tests on the new feature/area refactored will ease the review process and give more confidence when syncing the branch with master and solving conflicts.
### releasing the next version

If someone wants to test Coral with different versions of PHP, MariaDB, MySQL, it's gonna be tedious.

Same when regressions will be discovered and fixes are added. Someone will have to test again many stuff to be sure that the fixes didn't break something else.

Same for the person testing that the upgrade from the previous release works.
# How to test?

Here are diagrams that represent the situation and the possible testing strategy.

**A→B means that A depends on B**.
![diagram-plan-for-tests_300_dpi](https://cloud.githubusercontent.com/assets/2678215/17961518/89049f92-6aac-11e6-8b20-9f8ab9ae789a.png)
[source (ODG format)](https://github.com/Coral-erm/Coral/files/437356/plan-for-tests-diagram.zip)

To sum up this potentially confusing diagram: We can't start doing tests below the HTTP request level (except for maybe some small areas of the code), let's not even think of unit tests. Because the code is currently not testable.

GUI tests have their own long term issues, they don't scale well.
- testing multiple variations of a same scenario will be slow to write
- a comprehensive GUI test suite will be slow to run so it can't be ran every few minutes when developing
- maintenance cost when changing the GUI

Note that these issues are _relative_ to low level tests. The gap between no tests/manual tests and automated GUI test is much much bigger.

So the safest incremental approach would be to start doing GUI tests in order to get coverage. Because even if GUI tests aren't perfect, they are **way better than no tests at all** because right now our only coverage are manual GUI tests done by few people mostly at the end of a release cycle. So basically the first step is to automate them and allowing anyone to check that the most important features still work will already be a huge improvement in terms of QA and refactoring possibilities.
# Concretely how do we start testing Coral?

There are two approaches that will enable us to move toward a comprehensive test coverage for Coral
## 1. Add a test when working on a new feature, doing refactoring or a bugfix.

For example when extracting the views to a separate file, we would make a test about the content of the page before doing the refactoring. Thus allowing to use a part of the time that will be spent anyway in manually testing that nothing broke into rather building a test to guarantee that it will still work in the future.
## 2. Add tests for our most important processes

Instead of doing them manually.
## Choosing the tools

Before making a test we need to choose our tools.
### Criteria
- easy to setup and run
  The tools shouldn't require installing many external tools
- easy to use
  The tools should allow use to quickly write reliable tests even without requiring much testing experience
### Strategy

Due to my lack of experience in PHP testing tools, the approach was to make various proofs of concept with various testing tools and approaches. The initial set was chosen after asking and discussing at various moment on the IRC channels #php@freenode and #programming@freenode to know what was the "state of the art" in PHP testing and how to deal with legacy apps.
# Proofs of concept

The test in itself was simply to create a resource and delete it. As a consequence, login in is also required.

To try running the test, checkout on my different branches.

`git remote add tuxayo https://github.com/tuxayo/Coral.git`

`git fetch tuxayo`

`git checkout BRANCH_NAME`

Branches:

`experimental-php-unit-browser-emulator`

`experimental-php-unit-real-browser`

`experimental-codeception`
## PHPUnit + browser emulator (wrapper for HTTP request + HTML parser, not a real browser)

[See the code](https://github.com/Coral-erm/Coral/compare/master...tuxayo:experimental-php-unit-browser-emulator)

`git checkout experimental-php-unit-browser-emulator`

See [TESTS.md](https://github.com/tuxayo/Coral/blob/experimental-php-unit-browser-emulator/TESTS.md) for setup and usage instructions + details about the components involved.
### Pros
- supports low level tests (unit, integration, etc)
- fast runtime
- high level functions to browse, click, fill the forms and make assertion about the page
  Simpler and faster to write than manual requests + HTML parsing (that was horrible)
### Cons
- Most features are a bit hard to test because of heavy use of Ajax
  e.g. seeing the lists/indexes. Because we have to see the source code or use the browser network tool to know which url and which parms allow to see stuff.
- Many features are **extremely hard** to test
  - forms: almost all are triggered by Ajax to perform some validation
    - e.g. with resources it's not possible to use the browser network tool to easily capture the params due to the manual redirection made by the JS.
      So the other way is to inspect the JS and HTML to find what will be the request. Something went wrong and I managed to create a resource without any data and it's page is empty and the not even the delete button show up...
    - A lot of time(one have to always take into account what is Ajax and what is not)
    - Not very reliable results because one can forget to add params in forms that the front end actually enforces → can do things impossible in a browser and break Coral.
    - The test don't really show what a user it doing as there are steps to compensate the lack of JS everywhere. (raw requests, requests to Ajax content)
- Impossible to test many things
  - It can't be useful to cover the views enough to refactor/separate them from the logic.
- no Ajax, so we have to use the Ajax requests URLs directly.
  - Cost in time to pay attention of was is Ajax and what is not.
  - Can't just navigate as if one was on a real browser, so it takes more steps to see the data we want.
- HTML not the same because a real browser fixes it.
  Thus it may(e.g. happens with tables) will add tags (e.g. tbody) and CSS selectors which work on a real browser might not in the emulator.
## PHPUnit + Mink + real browser (PhantomJS)

PhantomJS: headless browser(no GUI) but very accurate page rendering, because it uses QTWebKit

Mink: Lib to interact with a browser (click, fill form, get content, etc)

[See the code](https://github.com/Coral-erm/Coral/compare/master...tuxayo:experimental-php-unit-real-browser)

`git checkout experimental-php-unit-real-browser`

See [TESTS.md](https://github.com/tuxayo/Coral/blob/experimental-php-unit-real-browser/TESTS.md) for setup and usage instructions + details about the components involved.
#### Same behavior as the real app

Like for a real real user, same HTML, Ajax works
#### Can basically test anything

Even bugs like https://github.com/Coral-erm/Coral/issues/73
Because if the HTML is broken enough so that the error message is not visible, the test will fail.
While a browser emulator will find the text of the error message and consider that it's here.
#### Allows for a decent code coverage

Compared to a browser emulator it will miss many server side code if one doesn't do many optional Ajax calls manually.
#### Doesn't automatically wait for page having loaded

Which makes hard to deal with Ajax.
I managed to make a helper function that works quite well to wait for the page to be ready but there are still some edge cases that will cause random failures. And one have to think to use the helper function for most of the pages.
When changing the language in the test, before login in, it took some time to make it work required some ugly (and potentially unreliable) sleep.
Codeception handled that seamlessly.
#### Slow runtime

Having dozens of tests will take several minutes to run
#### Issue with our strange menu links

Browsers dev tools show that the link itself doesn't have an area.
That's why it's not directly clickable with Mink (the lib in the PHPUnit + real browser PoC) and one must click on the inner div. (with the CSS selector provided to the click function)
## Codeception + real browser (PhantomJS)

[See the code with config and documentation](https://github.com/tuxayo/Coral/commit/experimental-codeception) or [only the test file](https://github.com/tuxayo/Coral/blob/experimental-codeception/tests/acceptance/ResourcesCept.php)

`git checkout experimental-codeception`

See [TESTS.md](https://github.com/tuxayo/Coral/blob/experimental-codeception/TESTS.md)  for setup and usage instructions + details about the components involved.
#### Wait automatically for the page+Ajax to finish loading

In almost all the cases where one would manually wait otherwise.
#### One can easily define helpers for all tests to share code and have cleaner tests

Helpers are just functions to help doing some task.
For tests it would be for example combining things that often come together.
Like waiting for the Ajax to finish before filling a form.
#### Handles without issue our strange menu links

Browsers dev tools show that the link itself doesn't have an area.
#### Less prone to randomly failing tests (flaky tests)

The final version of the test has been ran over 9000 times without issues. (except for something strange, see cons)
It's usually a bit hard avoid having flaky tests but it seems that Codeception manages to handle for us most of the cases where one would have to do a specific check like `waitForText` or `waitForElement`.
#### Displays nicely the tests steps

Useful to debug or as a **documentation** to learn some complex features.

With the option `-vv`

![](https://cloud.githubusercontent.com/assets/2678215/18161613/e9c4044a-7032-11e6-8393-698f45f2fa28.png)
#### Slow runtime

Having dozens of tests will take several minutes to run.
#### PhantomJS can freeze a some times

During all the days of working on that stack, few times PhantomJS ended up waiting indefinitely for the page to load.
The workaround is to relaunch PhantomJS and the test run (which would have failed)
It seems to be a known bug in PhantomJS which hopefully will get fixed.
On over 30000 test runs on that stack, this issue happened a dozen of times and there were three series of 3100, 6000 and 9100 consecutive runs that worked without any issue (running them in a loop is a good way to check their robustness)

From what I recall it didn't happened with PHPUnit + Mink + PhantomJS but as it can't run a few dozens times in a row without failing due to it's own issues, I couldn't see if the PhantomJS issue happens because usually it happens after a few hundred runs with Codeception + PhantomJS.

So it's not a blocking issue as we only need to run the test once. Running them many times is useful to detect flaky tests.

**edit:** It _seems_ that it happens when loading external resources like `font-awesome.min.css` and `https://fonts.googleapis.com/css?family=Open+Sans:100,400,300,600,700` which could be embedded in Coral so this issue should be mitigable.
# Other tools/techniques considered (skip if you don't have much time)
## Why not just HTTP requests tests ?

It was tried but:
- the server responds HTML even in Ajax requests so I need to parse HTML to check the response and the most efficient tools (and easier to use) are the GUI test tools
- they don't cover the JS which contains a decent part of the logic (and most user input validation)
- they don't cover that the HTML is not totally broken. A lot of refactoring is needed to separate the logic and the view so regressions on that side are to be expected.
- the aim is to automate the manual checking (installer, main features) that is systematically done and consuming our low QA resources (delaying releases and pull request merges)
## BDD syntax (Gherkin)

Doable with Codeception and Behat.
If we don't have non coders writing tests, then it might not be that useful.
Except if the other devs especially like the principle but it's more likely that this will be just another layer of abstraction to learn without obvious benefits.
## PhpSpec
- Only BDD style tests. It might not be wise to choose only this style.
- Smaller userbase: less doc, answers to questions and future-proofness
#### Behat
- Only BDD style tests. It might not be wise to choose only this style.
- Smaller userbase: less doc, answers to questions and future-proofness
# References (skip if you don't have much time or are not curious)
## The talk [Test-Driven Repair](https://www.youtube.com/watch?v=1i5leCslA4g) by Christopher Neugebaue

40min length

It sums up a lot of the though process and the conclusions about how to start testing Coral. In the second half(from minute 23) it also addresses architectural strategies to move towards from a legacy app to more maintainable and easily testable (low level tests) one.
## The talk "Inverting The Testing Pyramid" by Naresh Jain

[Slides](http://www.slideshare.net/nashjain/inverting-the-testing-pyramid)

It(talk + slides) was a great help to understand the different kinds of tests and the long term issues with only GUI tests. It shows the benefits of continuous integration (very short lived branches among other things)
# Conclusion

Codeception + PhantomJS are the tools that are the simplest to setup and more importantly which are the most accessible to begin making tests without prior testing experience.

So let's start testing and make Coral rock? :)

PR #115 is the mergable version of the Codeception PoC.

**Comment by tuxayo** [comment]
With #130 (a follow up of #115) merged, it's done! :-D
